File: server/TWS Foundation/CSM Foundation/Database/Models/Options/Filters/SetViewDateFilter.cs
using System.Linq.Expressions;

using CSM_Foundation.Database.Interfaces;

namespace CSM_Foundation.Database.Models.Options.Filters;

/// <summary>
/// 
/// </summary>
/// <typeparam name="TSet"></typeparam>
public class SetViewDateFilter<TSet>
   : ISetViewFilter<TSet>
    where TSet : ISet {
    public static readonly string Discriminator = typeof(SetViewDateFilter<TSet>).Name;
    public string Discrimination { get; init; } = Discriminator;
    
    
    public string Property { get; set; } = "Timestamp";
    public int Order { get; set; }


    public required DateTime From { get; set; }
    public DateTime? To { get; set; }

    public Expression<Func<TSet, bool>> Compose() {
        ParameterExpression param = Expression.Parameter(typeof(TSet), "X");
        MemberExpression prop = Expression.PropertyOrField(param, Property);

        ConstantExpression fromConstant = Expression.Constant(From, typeof(DateTime));
        BinaryExpression fromEvaluation = Expression.GreaterThanOrEqual(prop, fromConstant);

        BinaryExpression expression = fromEvaluation;

        if (To != null) {
            ConstantExpression toConstant = Expression.Constant(To, typeof(DateTime));
            BinaryExpression toEvaluation = Expression.LessThanOrEqual(prop, toConstant);
            expression = Expression.AndAlso(fromEvaluation, toEvaluation);
        }

        return Expression.Lambda<Func<TSet, bool>>(expression, param);
    }
}

File: client/tws_foundation_client/lib/src/models/options/filters/set_view_date_filter.dart
import 'package:csm_client/csm_client.dart';
import 'package:tws_foundation_client/src/models/interfaces/set_view_filter_interface.dart';

final class SetViewDateFilter<TSet extends CSMSetInterface> implements SetViewFilterInterface<TSet> {
  @override
  final String discrimination = 'SetViewDateFilter`1';

  ///
  @override
  final String property;

  ///
  @override
  final int order;

  final DateTime from;

  final DateTime? to;

  ///
  const SetViewDateFilter(
    this.order,
    this.from,
    this.to, {
    this.property = 'Timestamp',
  });

  @override
  JObject encode() {
    return <String, dynamic>{
      'discrimination': discrimination,
      'property': property,
      'order': order,
      'from': from.toIso8601String(),
      'to': to?.toIso8601String(),
    };
  }
}
